grpc-ingest/src/grpc.rs

```diff
@@ -109,107 +110,227 @@ impl<'a> AsyncHandler<GrpcJob, topograph::executor::Handle<'a, GrpcJob, Nonblock
                 }
             }
 
-            grpc_tasks_total_dec();
+            grpc_tasks_total_dec(&label, &stream_config.name.to_string());
         }
     }
 }
 
 pub async fn run(config: ConfigGrpc) -> anyhow::Result<()> {
     let redis_client = redis::Client::open(config.redis.url.clone())?;
-    let config = Arc::new(config);
     let connection = redis_client.get_multiplexed_tokio_connection().await?;
 
     let mut shutdown = create_shutdown()?;
 
-    let pipe = Arc::new(Mutex::new(TrackedPipeline::default()));
-
-    let mut accounts = HashMap::with_capacity(1);
-    let mut transactions = HashMap::with_capacity(1);
-
-    accounts.insert("das".to_string(), config.accounts.filter.clone().to_proto());
-    transactions.insert(
-        "das".to_string(),
-        config.transactions.filter.clone().to_proto(),
-    );
-
-    let request = SubscribeRequest {
-        accounts,
-        transactions,
-        ..Default::default()
-    };
-
-    let mut dragon_mouth_client =
-        GeyserGrpcClient::build_from_shared(config.geyser_endpoint.clone())?
-            .x_token(config.x_token.clone())?
-            .connect_timeout(Duration::from_secs(10))
-            .timeout(Duration::from_secs(10))
-            .connect()
-            .await
-            .context("failed to connect to gRPC")?;
-
-    let (mut subscribe_tx, stream) = dragon_mouth_client
-        .subscribe_with_request(Some(request))
+    let subscription_tasks = SubscriptionTask::build()
+        .config(config)
+        .pipeline(TrackedPipeline::default())
+        .connection(connection)
+        .start()
```

Comment:
The code should be easier to reason about if subscription task manages a single subscription and here you do the loop and instantiate n instance of subscription task.

This would also allow tracked pipeline to be for a single subscription as well.

Reply:
separated to handle each subscription now